### ISSUE TITLE
Bazel is stuck fetching external repositories

### ISSUE THREAD

**Issue by luispadron** [bazelbuild/bazel#21438]
### Description of the bug:

We have reports of Bazel consistently hanging during fetching of external repositories. Particular, those consumed with `git_override` / `git_repository`. 

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

I do not have a reproducible example or a way to consistently reproduce this. When this happens it's been reported that it'll consistently hang even on server restarts. The only thing that helps is restarting the users machine until they eventually hit it again.

### Which operating system are you running Bazel on?

macOS 14

### What is the output of `bazel info release`?

release 6.5.0

### Have you found anything relevant by searching the web?

I've found some other threads in Slack and issues on GitHub related to hanging but nothing that seems like this.

### Any other information, logs, or outputs that you want to share?

Some information about our environment:

- We are using arm64 machines
- Xcode 15.1 toolchains
- Bzlmod enabled (without lockfile)

Our reporter also managed collect 3 jstack traces when the Bazel command was hung. I've attached those below.
[bazel_jstack_trace.log](https://github.com/bazelbuild/bazel/files/14350796/bazel_jstack_trace.log)

**Comment by Wyverald** [comment]
just from a brief look at the 3 stack traces, they seem to be hung on different things. I couldn't immediately tell what the first two are about, but the third one seems to be waiting for the `git` subprocess to exit. Is the `git` process still alive when you took the stack trace?

**Comment by luispadron** [comment]
@Wyverald thanks for the fast reply!

Yeah, as far as we can till `git` was still alive. Any thoughts on how to debug further? For the person its happening to it happens on a daily basis but we do not have a lot of reports of this exact issue

**Comment by Wyverald** [comment]
since it sounds like this is some problem with `git`, I'm going to go ahead and close this. Feel free to reopen if this is actually Bazel-related.

**Comment by luispadron** [comment]
Makes sense, thank you!

**Comment by efirestone** [comment]
Slight clarification: when I killed the `git` process that was accessing `/Users/USER/Development/<repo>/.git/fsmonitor--daemon.ipc`, it unblocks the Bazel build. That process respawns, but it takes a bit and in the meantime seems to unblock Bazel's `git` invocations. Killing it also likely frees up a bunch of FS event openings.

I'll try `git config --global core.fsmonitor false` and see if that helps with things in the long run.

Thanks for your time, and the pointer towards looking at the `git` processes.

**Comment by cgrindel** [comment]
I just started seeing this with the `tca_example` in `rules_swift_package_manager`.

@luispadron Did you ever come up with fix or workaround?

**Comment by luispadron** [comment]
@cgrindel just what Eric posted above, disabling the git fsmonitor to fix this. 

It doesn't seem like we think this is a Bazel issue but maybe some more investigation needs to happen here?

**Comment by cgrindel** [comment]
Thanks for the reply. I also found [this discussion](https://bazelbuild.slack.com/archives/C014RARENH0/p1711487321624049). I am trying the [--experimental_worker_for_repo_fetching](https://bazel.build/reference/command-line-reference#flag--experimental_worker_for_repo_fetching) set to `off`.

**Comment by luispadron** [comment]
@Wyverald Would it make sense to update `git_worker` to use `"-c", "core.fsmonitor=false"` so that we do not run into this issue when fetching many repositories.

Tools like Swift Package Manager do something similar: https://github.com/swiftlang/swift-package-manager/blob/main/Sources/SourceControl/GitRepository.swift#L157-L173

I can PR this if we think it makes sense. This issue is closed but several of us are still hitting this with the only solution we've found being to disable the global fsmonitor setting which is less ideal.

**Comment by Wyverald** [comment]
Some light code archeology seems to indicate that this flag was added to fix a Windows-specific for Swift: https://github.com/swiftlang/swift-package-manager/pull/4032

Do you have an idea of what other consequences of adding this flag could be? I personally have no idea what the global fsmonitor even is, so some context would help with the decision.

**Comment by luispadron** [comment]
So what we think is happening is that when using rules like rules_swift_package_manager that use the Bazel git_worker to fetch a large amount of git repositories the OS is running out of file watchers and thus blocking during repository fetch. This has happened to us in Square and Cash App as we've started to move more things to rules_swift_package_manager. 

git by default seems to spin up a file system monitor during each clone which can exhaust the limit available on a given machine if there are many parallel or subsequent clones. 

We've worked around this by disabling file system monitoring in git globally, but a potentially more scoped approach of disabling the fsmonitor only on the repositories being fetched by Bazel would also work and be more ideal. This would in theory allow us to continue monitoring our own local repositories

**Comment by Wyverald** [comment]
Thanks for the context.

> git by default seems to spin up a file system monitor during each clone which can exhaust the limit available on a given machine if there are many parallel or subsequent clones.

If this is the case, I'm happy to accept a PR. We can run some downstream testing to see if it breaks anyone.

**Comment by luispadron** [comment]
Great I'll open a PR soon, should we reopen this issue and have the eventual PR close it instead, for tracking purposes.

**Comment by luispadron** [comment]
So I believe (with the testing I've done) that https://github.com/bazelbuild/bazel/pull/23626 solves this issue within `git_worker`.

I performed tests by spawning many file watchers using the `fswatch` tool, using a released Bazel version `7.3.1` when running `bazel fetch` in the `examples/lots_of_repos` directory of the branch: https://github.com/cgrindel/rules_swift_package_manager/tree/luis/lots-of-repo-example I was able to consistently reproduce stuck fetches. 

I confirmed that building `//src:bazel` and using that instead consistently allows all fetches to continue and I ran into no hangs during the process.

---

```sh
#!/bin/bash

# Number of monitors to create
NUM_MONITORS=100  # Adjust based on your stress test needs

# Directory to monitor (can be the same or different for each monitor)
MONITOR_DIR="/tmp/fsmonitor-test"

# Create the directory if it doesn't exist
mkdir -p "$MONITOR_DIR"

# Array to store process IDs of background fswatch processes
declare -a PIDS

# Function to create and run a file system monitor
create_monitor() {
  local dir=$1
  fswatch "$dir" >/dev/null &
  PIDS+=($!)  # Store the process ID
}

# Create the specified number of monitors
for i in $(seq 1 $NUM_MONITORS); do
  create_monitor "$MONITOR_DIR"
  
  # Optionally, introduce a slight delay to avoid overwhelming the system too quickly
  sleep 0.1
done

echo "Created $NUM_MONITORS file system monitors."


# Function to clean up by killing all background processes
cleanup() {
  echo "Cleaning up monitors..."
  for pid in "${PIDS[@]}"; do
    kill "$pid" 2>/dev/null
  done
  echo "All monitors terminated."
}

# Trap script exit (Ctrl+C or normal exit) to ensure cleanup happens
trap cleanup EXIT

# Keep the script running to simulate long-running monitors
echo "Press Ctrl+C to stop the script and clean up the monitors."
while true; do
  sleep 10  # Keep the script alive indefinitely
done

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.4.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc1`. Thanks!
